### PR TITLE
feat: Refactor to make ETL classes more modularised

### PR DESCRIPTION
**Pull request by arihantsurana**
### Link to Github Issue
---
* https://github.com/hipagesgroup/data-tools/issues/119
* https://github.com/hipagesgroup/data-tools/issues/120

### What changes are proposed in this PR?
---
* Added some abstract base classes to serve as the foundation of a more modular and extensible ETL framework
* Refactored S3ToS3 etl to use the new framework
* Added example

### How was this tested?
---
* Unit tests

### CODE REVIEW

File: hip_data_tools/etl/s3.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+    bucket: S3Bucket
+    key_prefix: S3Key
+    suffix: Optional[str]
+    connection_settings: AwsConnectionSettings
+
+
+class S3Extractor(Extractor):
+    def __init__(self, settings: S3SourceSettings):
+        super().__init__(settings)
+        self.settings = settings
```
Comment (vbmudalige) [inline hip_data_tools/etl/s3.py:28]: shouldn't this be private ? 
`self.__settings`

File: hip_data_tools/etl/s3.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+        pass
+
+
+class S3FilesExtractor(S3Extractor):
+    def __init__(self, settings: S3SourceSettings):
+        super().__init__(settings)
+        self._source_keys = None
+        self.file_counter = 0
+
+    def list_source_files(self) -> List[S3Key]:
```
Comment (vbmudalige) [inline hip_data_tools/etl/s3.py:54]: Nice to have some doc comments for public methods

File: example/etl/s3_to_s3_file_tansfer.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+#  If you want to use the Aws profiles stored by the swd cli tool, uncomment the following code:
+# aws_setting = AwsConnectionSettings(
+#     region="us-east-1",
+#     secrets_manager=None,
+#     profile="default")
+
+# Define the ETL instance
+etl = S3ToS3FileCopy(
+    source=s3.S3SourceSettings(
```
[the remark this answers is not in the document]
Comment (Gurado) [reply example/etl/s3_to_s3_file_tansfer.py:24]: this sounds concerning @arihantsurana

File: hip_data_tools/etl/s3.py [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+    types of data from these files
+
+    Args:
+        settings (S3SourceSettings):  settings to connect to the source s3 bucket
+    """
+
+    def __init__(self, settings: S3SourceSettings):
+        super().__init__(settings)
+        self._source_keys = None
+        self.file_counter = 0
```
Comment (vbmudalige) [inline hip_data_tools/etl/s3.py:73]: We can make file_counter private

**Pull request merged**